### ISSUE TITLE
Ability to have `uuid` for the primary key?

### ISSUE THREAD

**Issue by mmahalwy** [stephenh/joist-ts#192]

**Comment by stephenh** [comment]
This seems doable with probably not too many gyrations; are you thinking db-assigned uuids or client-assigned uuids?

If client-assigned uuids, the `id` params could be made required, so clients have to pass them when creating.

And the "assign ids" step in `PostgresDriver` would have to be skipped, based on a config flag of some sort.

Both code generation (i.e. via `joist-codegen.json` flag) and runtime (i.e. `PostgresDriver` via a constructor arg) would probably need to be told to use client-assigned ids.

Fwiw I am currently pretty sold on keeping the requirement for a single primary key field, and that the name of that field is `id`.

**Comment by mmahalwy** [comment]
Hey - trying to find where we can do `db-assigned` uuids here. Is it just on migrations?

**Comment by stephenh** [comment]
@mmahalwy what does the DDL of your table look like (just the `id` column)? Is it like this:

```
    contact_id uuid DEFAULT uuid_generate_v4 (),
```

From https://www.postgresqltutorial.com/postgresql-uuid/

**Comment by mmahalwy** [comment]
```sql
CREATE TABLE "Application" (
    "id" UUID NOT NULL DEFAULT gen_random_uuid(),
)
```

**Comment by mmahalwy** [comment]
We are also migrating this first:
```sql
CREATE EXTENSION IF NOT EXISTS pgcrypto;
```

I know there's also:
```sql
CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
````

**Comment by mmahalwy** [comment]
@stephenh sorry I missed your last comment. I see – yeah the driver would need to be changed to support built-in UUID generation. We can have it be done in the driver to 'fake' it, as you've suggested, but I'd imagine folks may want this to be natively supported as joist may not be the only thing touching the db – just like folks want created at and updated at to auto populate.

**Comment by stephenh** [comment]
> I'd imagine folks may want this to be natively

Ah yeah, agreed; my assumption is that the db could have an `id` with a default uuid (for non-Joist / raw SQL clients), but Joist could still generate `INSERT`s with its own Joist/JS-generated uuids, and so be able to still up-front assign ids.

**Comment by stephenh** [comment]
I've start poking around at this in #223 . Besides just the feature itself, I'm also feeling out how to have a sane repository / test case setup, i.e. I'm trying out a new `uuid-ids` package that would be separate from the current large/covers-all-the-things `integration-tests` package.

So far the only thing I've done is add `uuid` as a database type, so that `npm run codegen` doesn't immediately blow up, but now there are compile errors (which is somewhat expected, I haven't changed anything in the codegen yet). So, dunno, probably ~4-5 rounds of "run it and see what breaks" and I'll keep seeing how it goes.

**Comment by stephenh** [comment]
@mmahalwy I merged an initial version of this in #223; see the `packages/tests/uuid-ids` mini-project for an example of how the basics work. Note that I've only tested loading and savings, i.e. I've not tried queries yet (they should? work) and have not like ran the whole joist test suite against the "with uuid ids" codepaths (although I don't think that'd be necessary, but just fwiw / as a disclaimer).

Feel free to file any follow up issues around `uuid` keys, but I'll close this initial issue as fixed. Thanks!
